counsel.el

```diff
@@ -2245,18 +2245,50 @@ regex string."
 (ivy-set-occur 'counsel-ag 'counsel-ag-occur)
 (ivy-set-display-transformer 'counsel-ag 'counsel-git-grep-transformer)
 
+(defconst user-input-switch "-- ")
+(defconst user-input-switch-length (length user-input-switch))
+
+(defun parse-user-input (string)
+  "Parse user input in STRING into a pair of two strings: (query extra-ag-args).
+Where query is the search term and extra-ag-args are any additional switches."
+  (let ((query string)
+        (extra-ag-args ""))
+    (when (string-prefix-p "-" string)
+      (let ((index (string-match user-input-switch string)))
+       (when index
+          (setq query (substring string (+ user-input-switch-length index)))
+          (setq extra-ag-args (substring string 0 index)))))
+    (cons query extra-ag-args)))
+
+(defun format-counsel-ag-command (extra-ag-args term)
+  "Format counsel-ag-command as a string.
+EXTRA-AG-ARGS is a string of the additional arguments.
+TERM is the search string."
+  (let ((args-end (string-match " -- " extra-ag-args))
+        (file ""))
+    (when args-end
+      (setq file (substring-no-properties extra-ag-args (+ args-end 3)))
+      (setq extra-ag-args (substring-no-properties extra-ag-args 0 args-end)))
+    (format counsel-ag-command
+            (concat extra-ag-args
+                    term
+                    file))))
```

Comment:
I think this can be simplified a bit:

```el
(defun counsel--format-ag-command (extra-args needle)
  "Construct a complete `counsel-ag-command' as a string.
EXTRA-ARGS is a string of the additional arguments.
NEEDLE is the search string."
  (format counsel-ag-command
          (if (string-match " \\(--\\) " extra-args)
              (replace-match needle t t extra-args 1)
            (concat extra-args needle))))
```

100% guaranteed untested.

Either way, I think the function should be redesigned or at least documented better so that it is clearer whether `extra-ag-args`/`term` should contain leading or trailing spaces etc., because at the moment, if there is no ` -- ` present, `term` just gets appended to `extra-ag-args` without a space in-between.

Reply:
@DEADB17 WDYT?

Reply:
Sorry, I won't be able to get back to this until I'm back home tonight.

Reply:
I tested your version and didn't find any issues. Thanks.
I also added an extra space when concatenating `(concat extra-args " " needle)` so that callers don't have to care about it.
I also tried trimming the arguments with `string-trim` before `concat`ing them but since white-space makes no difference when calling the command I decided to remove it in the end and avoid the extra work.